src/datachain/catalog/catalog.py

```diff
@@ -1506,13 +1549,45 @@ def query(
         if capture_output:
             popen_kwargs = {"stdout": subprocess.PIPE, "stderr": subprocess.STDOUT}
 
+        def raise_termination_signal(sig: int, _: Any) -> NoReturn:
+            raise TerminationSignal(sig)
+
         with subprocess.Popen(cmd, env=env, **popen_kwargs) as proc:  # noqa: S603
-            if capture_output:
-                args = (proc.stdout, output_hook)
-                thread = Thread(target=_process_stream, args=args, daemon=True)
-                thread.start()
-                thread.join()  # wait for the reader thread
+            orig_sigint_handler = signal.getsignal(signal.SIGINT)
+            # ignore SIGINT in the main process
+            signal.signal(signal.SIGINT, signal.SIG_IGN)
```

Comment:
We need two different behaviors from this `query` API that this has become complicated.

CLI should ideally ignore `SIGINT` and forward `SIGTERM` to the child.
However, Celery tasks should not ignore `SIGINT`, but consider `SIGTERM` as a special signal for graceful termination.

Here, I have made a compromise: to ignore `SIGINT` and use `SIGTERM` as a special signal for graceful termination everywhere. 

So, `SIGINT` still works for CLI (where it matters most), and `SIGTERM` will work for canceling celery tasks.

Reply:
so, I'm a bit lost - are we ignoring SIGINT in CLI or not?

Reply:
We are ignoring `Ctrl-C` in the main process. If you are running in terminal, OS (`tty` driver) will send the signal to all processes in the foreground process group, so the script will also get the signal. If we forward signals, then the subprocess will receive the signal twice.

If you are running a background process (e.g.: via a script), you'd have to interrupt the script subprocess rather than the CLI process.

Reply:
> will send the signal to all processes in the foreground process group

okay, I see

I think it would be helpful to describe these details in the comment for the next generations, it's a dense, low level code